### PR TITLE
chore: integrate sqlconnect-go for snowflake

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Integrate sqlconnect-go for snowflake

## Linear Ticket

- Resolves PIPE-1343

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -80,7 +80,7 @@ require (
 	github.com/rudderlabs/rudder-observability-kit v0.0.3
 	github.com/rudderlabs/rudder-schemas v0.5.0
 	github.com/rudderlabs/sql-tunnels v0.1.7
-	github.com/rudderlabs/sqlconnect-go v1.7.1
+	github.com/rudderlabs/sqlconnect-go v1.7.2-0.20240725064850-4e9872d5ae95
```
Comment (ktgowtham) [inline go.mod:83]: just curious, how's the package with this version created?

**Review by ktgowtham — APPROVED** [review]
LGTM

File: go.mod (hunk 1 above)
Comment (achettyiitr) [reply go.mod:83]: Was waiting for the [1.7.2](https://github.com/rudderlabs/sqlconnect-go/pull/145) to be released. Will update the PR once it is merged.

Comment (achettyiitr) [reply go.mod:83]: https://www.notion.so/rudderstacks/Working-with-multiple-modules-804e0fd7374e4e89a84af64ec9fd170b?pvs=4#5442e7e7ce144a81a55eb92b18392013

File: warehouse/integrations/snowflake/snowflake.go [hunk 2]
```diff
@@ -1062,50 +1047,52 @@ func (sf *Snowflake) LoadUserTables(ctx context.Context) map[string]error {
...
-	if db, err = sql.Open("snowflake", dsn); err != nil {
-		return nil, fmt.Errorf("SF: snowflake connect error : (%v)", err)
+	sqlConnectDB, err := sqlconnect.NewDB("snowflake", credentialsJSON)
+	if err != nil {
+		return nil, fmt.Errorf("creating sqlconnect db: %w", err)
 	}
 
 	alterStatement := `ALTER SESSION SET ABORT_DETACHED_QUERY=TRUE`
-	_, err = db.ExecContext(ctx, alterStatement)
+	_, err = sqlConnectDB.SqlDB().ExecContext(ctx, alterStatement)
```
Comment (cisse21) [inline warehouse/integrations/snowflake/snowflake.go:1090]:
```suggestion
	sqlDB := sqlConnectDB.SqlDB()
	_, err = sqlDB.ExecContext(ctx, alterStatement)
```

File: warehouse/integrations/snowflake/snowflake.go [hunk 3]
```diff
@@ -1062,50 +1047,52 @@ func (sf *Snowflake) LoadUserTables(ctx context.Context) map[string]error {
...
 	if err != nil {
 		return nil, fmt.Errorf("SF: snowflake alter session error : (%v)", err)
 	}
 	middleware := sqlmw.New(
-		db,
+		sqlConnectDB.SqlDB(),
```
Comment (cisse21) [inline warehouse/integrations/snowflake/snowflake.go:1095]:
```suggestion
		sqlDB,
```

**Review by cisse21 — APPROVED** [review]

File: warehouse/integrations/snowflake/snowflake.go [hunk 4]
```diff
@@ -1062,50 +1047,52 @@ func (sf *Snowflake) LoadUserTables(ctx context.Context) map[string]error {
...
-		}
-
-		urlConfig.PrivateKey = rsaPrivateKey
-		urlConfig.Authenticator = snowflake.AuthTypeJwt
-	} else {
-		urlConfig.Password = cred.password
-		urlConfig.Authenticator = snowflake.AuthTypeSnowflake
-	}
+	var (
+		account              = whutils.GetConfigValue(configKeyAccount, sf.Warehouse)
```
Comment (lvrach) [inline warehouse/integrations/snowflake/snowflake.go:1051]: Not introduced here, but I noticed how GetConfigValue works.

1. Do we really need: `	configKey := fmt.Sprintf("Warehouse.pipeline.%s.%s.%s", warehouse.Source.ID, warehouse.Destination.ID, key)`
2. Can we replace it with a non global config method?

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:1051]: Regarding 1, this was added as providing pipeline level sync frequency override in [here](https://github.com/rudderlabs/rudder-server/pull/3094).

Regarding 2, Yeah we can do that but that would require passing *`config.Config` as a parameter, something like this:
```GO
func GetConfigValue(conf *config.Config, key string, warehouse model.Warehouse) (val string) {
	configKey := fmt.Sprintf("Warehouse.pipeline.%s.%s.%s", warehouse.Source.ID, warehouse.Destination.ID, key)
	if conf.IsSet(configKey) {
		return conf.GetString(configKey, "")
	}
	destConfig := warehouse.Destination.Config
	if destConfig[key] != nil {
		val, _ = destConfig[key].(string)
	}
	return val
}
```

Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:1051]: Regarding 2, I can take it as part of a separate task.

File: go.mod (hunk 1 above)
Comment (achettyiitr) [reply go.mod:83]: Based on the commit (`4e9872d5ae956ee7fd8de27e28ecf123e8d836df`) I used, it modified the go.mod accordingly.

Comment (achettyiitr) [reply go.mod]: Upgraded to 1.7.3

File: warehouse/integrations/snowflake/snowflake.go (hunk 4 above)
Comment (achettyiitr) [reply warehouse/integrations/snowflake/snowflake.go:1051]: https://linear.app/rudderstack/issue/PIPE-1365/handle-review-comment-for-global-config

**Pull request merged**